primitives/npos-elections/src/pjr.rs

```diff
@@ -0,0 +1,519 @@
+ // This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implements functions and interfaces to check solutions for being t-PJR.
+//!
+//! PJR stands for proportional justified representation. PJR is an absolute measure to make
+//! sure an NPoS solution adheres to a minimum standard.
+//!
+//! See [`pjr_check`] which is the main entry point of the module.
+
+use crate::{
+	Candidate,
+	CandidatePtr,
+	Edge,
+	ExtendedBalance,
+	IdentifierT,
+	Support,
+	SupportMap,
+	Supports,
+	Voter,
+	VoteWeight,
+};
+use sp_std::{rc::Rc, vec::Vec};
+use sp_std::collections::btree_map::BTreeMap;
+use sp_arithmetic::{traits::Zero, Perbill};
+
+/// The type used as the threshold.
+///
+/// Just some reading sugar; Must always be same as [`ExtendedBalance`];
+type Threshold = ExtendedBalance;
+
+/// Compute the threshold corresponding to the standard PJR property
+///
+/// `t-PJR` checks can check PJR according to an arbitrary threshold. The threshold can be any value,
+/// but the property gets stronger as the threshold gets smaller. The strongest possible `t-PJR` property
+/// corresponds to `t == 0`.
+///
+/// However, standard PJR is less stringent than that. This function returns the threshold whose
+/// strength corresponds to the standard PJR property.
+///
+/// - `committee_size` is the number of winners of the election.
+/// - `weights` is an iterator of voter stakes. If the sum of stakes is already known,
+///   `std::iter::once(sum_of_stakes)` is appropriate here.
+pub fn standard_threshold(
+	committee_size: usize,
+	weights: impl IntoIterator<Item = ExtendedBalance>,
+) -> Threshold {
+	weights
+		.into_iter()
+		.fold(Threshold::zero(), |acc, elem| {
+			acc.saturating_add(elem)
+		})
+	/ committee_size as Threshold
```

Comment:
```suggestion
	/ committee_size.max(1) as Threshold
```

Overall, always remember that any code that will end up in the wasm runtime must never never ever panic (even in rare/mistake cases like this).

Reply:
(same reason why I asked for sum to change to fold)

Reply:
Good catch!